_drafts/device-logging.md

```diff
@@ -0,0 +1,273 @@
+---
+title: Trade-offs and considerations for logging on embedded devices
+description: "Trade-offs and considerations for logging on embedded hardware"
+author: victorlai
+---
+
+Logging on client devices comes with a host of interesting, unique challenges that aren't found on the server &mdash; devices are generally more memory, bandwidth, space, and CPU constrained than a server, and devices often take weeks, months, or years to update, rather than days or hours. This has interesting ramifications on how we log: we can't log as much data as we'd want to, logs have a more finite lifespan, and the format of logs are much more permanent, so if we don't log the right data in the right way to begin with, it can be months before we ship the next release with new logs. All that being said, there's also a lot of literature and learnings from logging on the server that we can also apply on the client side.
+
+In this article, I discuss some of the problems I’ve encountered in both producing and analyzing device logs over the years, on single digit number of devices up to hundred of thousands, and concepts I’ve applied to manage that complexity. My background is Android, so some of the examples may skew towards much more powerful hardware than simple microcontrollers.
+
+{% include newsletter.html %}
+
+{% include toc.html %}
+
+## Log collection
```

Comment:
I like it! (though I wish the sub sub headings were in the TOC too lol)

Reply:
I think sub sub headings might be clutter the TOC a bit from this snapshot view but it would technically be more complete. Probably more important for documentation to have a complete TOC than articles since we don't typically need good nav through an article